src/poly/cse.rs

```diff
@@ -1,134 +1,121 @@
-            if let Some(existing_expr) = existing_weak.upgrade() {
-                result.push(existing_expr);
-                continue;
+fn replace_common_subexprs_rec<
+    F: Field + Hash,
+    V: Clone + Eq + PartialEq + Hash + Debug,
+    SF: SignalFactory<V>,
+>(
+    decomp: &mut ConstrDecomp<F, V>,
+    constr: Expr<F, V, ()>,
+    common_ses: &[Expr<F, V, HashResult>],
+    assignments: &VarAssignments<F, V>,
+    signal_factory: &mut SF,
+) -> Expr<F, V, ()> {
+    // Check if the expression is a common subexpression
+    if let Some(_common_expr) = common_ses
+        .iter()
+        .find(|ce| constr.hash(assignments).meta().hash == ce.meta().hash)
+    {
+        // If it is, check if the signal already exists
+        let hash = constr.hash(assignments).meta().hash;
+        if let Some(signal) = decomp.auto_signals.iter().find_map(|(signal, expr)| {
+            if expr.hash(assignments).meta().hash == hash {
+                Some(signal.clone())
+            } else {
+                None
             }
+        }) {
+            Expr::Query(signal, ())
+        } else {
+            let new_signal = signal_factory.create("cse");
+            decomp.auto_eq(new_signal.clone(), constr);
+            Expr::Query(new_signal, ())
         }
-
-        // otherwise, insert the hash and the weak reference to the expression
-        seen_hashes.insert(hash, Rc::downgrade(&hashed_expr));
-        result.push(hashed_expr);
-    }
-
-    result
-}
-
-fn generate_random_assignment<F: Field + Hash, V: Debug + Clone + Eq + Hash>(
-    queriables: &Vec<V>,
-) -> VarAssignments<F, V> {
-    let mut rng = ChaCha20Rng::seed_from_u64(0);
-
-    let mut assignments = HashMap::new();
-
-    for queriable in queriables {
-        let value = F::random(&mut rng);
-        assignments.insert(queriable.clone(), value);
-    }
-
-    assignments
-}
-
-#[cfg(test)]
-mod test {
-    use super::*;
-    use std::collections::HashSet;
-
-    use halo2_proofs::halo2curves::bn256::Fr;
-
-    use crate::{
-        poly::{Expr::*, ToExpr},
-        sbpir::{query::Queriable, ForwardSignal, InternalSignal},
-    };
-
-    #[test]
-    fn test_generate_random_assignment() {
-        let internal = InternalSignal::new("internal");
-        let forward = ForwardSignal::new("forward");
-
-        let a: Queriable<Fr> = Queriable::Internal(internal);
-        let b: Queriable<Fr> = Queriable::Forward(forward, false);
-        let c: Queriable<Fr> = Queriable::Forward(forward, true);
-
-        let vars = vec![a, b, c];
-
-        let keys: HashSet<Queriable<Fr>> = vars.iter().cloned().collect();
-
-        let assignments: VarAssignments<Fr, Queriable<Fr>> = generate_random_assignment(&vars);
-
-        println!("Assignments: {:#?}", assignments);
-
-        for key in &keys {
-            assert!(assignments.contains_key(key));
+    } else {
+        // Otherwise, replace the subexpressions recursively
+        match constr {
```

Comment:
@rutefig perhaps we can abstract this out, with a method that recieves a lambda in Expr. Like expr.apply_subexpressions(|ses| ... )